README.md
# NAME

WebService::Geocodio - A Perl interface to Geocod.io

# VERSION

version 0.05

# SYNOPSIS

    use 5.014;
    use WebService::Geocodio;
    use WebService::Geocodio::Location;

    my $geo = WebService::Geocodio->new(
        api_key => $ENV{GEOCODIO_API_KEY}
    );

    # Wrigley Field
    my $loc = WebService::Geocodio::Location->new(
        number => 1060,
        postdirection => 'W',
        street => 'Addison',
        suffix => 'Street',
        city => 'Chicago',
        state => 'IL',
    );

    # Could add more than one thing here, even bare strings are OK
    # 20050 = zip code in Washington DC
    $geo->add_location($loc, '20050');

    $geo->add_field('timezone');

    # prints:
    # Chicago: 41.947205791667, -87.656316875, CST
    # Chicago: 41.947180613636, -87.657167363636, CST
    # Washington: 38.893311, -77.014647, EST
    map { say $_->city, ": ", $_->lat, ", ", $_->lng, ", " $_->fields->timezone->name } $geo->geocode();

# OVERVIEW

This module is a fairly thin wrapper around the [Geocod.io](http://geocod.io)
geocoding web service.  This service currently supports US and many Canadian based
addresses at the moment.  Both forward and reverse geocoding is supported.

In my testing, the service is somewhat finicky about how addresses are
presented and stored; please read the service API documentation thoroughly 
to make sure you're getting the best quality results from the service.

You will need to obtain a free API key to use this library.

All errors are fatal and reported by `confess`.  If you want more graceful
error handling, you might want to try using [Try::Tiny](https://metacpan.org/pod/Try::Tiny).

# ATTRIBUTES

## api\_key

This is the geocod.io API key. It is required.

## locations

The list of locations you want to geocode.  These can be bare strings (if you like) or
you can use a fancy object like [WebService::Geocodio::Location](https://metacpan.org/pod/WebService::Geocodio::Location) which will serialize
itself to JSON automatically.

## fields

You may request the following fields be included in the results:

- cd

    Congressional District (for the current Congress). This will return detailed
    biographical information about the Senators and Congressional representative
    from this district in the `current_legislators` field which is structured
    as an array where the member of Congress is first, followed by the Senators
    for the entire state.

    It's possible there may be overlaps in a given location, so this field is
    structured as an array since there may be 2 or more possible districts
    for a given location. See the official docs for more information.

- cd113

    Congressional District (for the 113th Congress which ran through 2015)

- cd114

    Congressional District (for the 114th Congress which ran through 2017)

- cd115

    Congressional District (for the 115th Congress which runs through 2019)

- stateleg

    The state legislative divisions for this location. The results include both
    House and Senate, unless the location is unicameral like Nebraska or Washington
    D.C., then only a senate result is given.

- timezone

    The timezone of this location, UTC offset and whether it observes daylight
    saving time.

- school

    The unified or elementary/secondary school district identifiers for this location.

- census

    Appends various U.S. Census statistical codes to your lookup. See the API docs
    for more information about these codes and how they can be used with other
    census data sets.

# METHODS

## add\_location

This method takes one or more locations and stores it in the locations attribute.

## show\_locations

Show the locations currently set for geocoding.

## clear\_locations

If you want to clear the current list of locations, use this method.

## add\_field

This method takes one or more fields to include in a result set. Valid fields are:

- cd (current Congress)
- cd113
- cd114
- cd115
- stateleg
- timezone
- school
- census

Fields that do not match these valid names are silently discarded.

## geocode

Send the current list of locations to the geocod.io service.

Returns undef if there are no locations stored.

In a list context, returns a list of [WebService::Geocodio::Location](https://metacpan.org/pod/WebService::Geocodio::Location) objects.
In a scalar context, returns an arrayref of [WebService::Geocodio::Location](https://metacpan.org/pod/WebService::Geocodio::Location)
objects. The list of objects is presented in descending order of accuracy.

## reverse\_geocode

Send the current list of latitude, longitude pairs to the geocod.io service.

Returns undef if there are no locations stored.

In a list context, returns a list of [WebService::Geocodio::Location](https://metacpan.org/pod/WebService::Geocodio::Location) objects.
In scalar context, returns an arrayref of [WebService::Geocodio::Location](https://metacpan.org/pod/WebService::Geocodio::Location) 
objects.  The list of objects is presented in descending order of accuracy.

# AUTHOR

Mark Allen <mrallen1@yahoo.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2018 by Mark Allen.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Change history for WebService-Geocodio

{{$NEXT}}

0.05        2018-02-18
	    Modernize dependencies
	    - Switch to Moo since strictures 2 does not fatalize all warnings
            Documentation change.
	    Use version 1.2 of the API
	    Support census field output
	    Support new congressional data output

0.04        2014-04-10
            Support fields as blessed objects rather than naked hashes.
            See API docs to see what fields and data are available from
            the service.

0.03        2014-03-25
            Major code refactor
            Add support for reverse geocoding
            Add rudimentary support for geocod.io fields

0.02        2014-01-23
            Fix testing failure when an api key isn't in the environment.
            There are *no* API changes.

0.01        2014-01-22
            Initial release


lib/WebService/Geocodio/Fields.pm
use strict;
use warnings;

{
    package WebService::Geocodio::Fields::CongressionalDistrict;

    use Moo;
    use strictures 2;;

    has [qw(name district_number congress_number proportion congress_years current_legislators)] => (
        is => 'ro',
    );

}

################## State Legislative Districts

{
    package WebService::Geocodio::Fields::StateLegislativeDistrict::House;

    use Moo;
    use strictures 2;;

    has [qw(name district_number)] => (
        is => 'ro',
    );
}

{
    package WebService::Geocodio::Fields::StateLegislativeDistrict::Senate;

    use Moo;
    use strictures 2;;

    has [qw(name district_number)] => (
        is => 'ro',
    );
}

{

    package WebService::Geocodio::Fields::StateLegislativeDistrict;

    use WebService::Geocodio::Fields::StateLegislativeDistrict::House;
    use WebService::Geocodio::Fields::StateLegislativeDistrict::Senate;

    use Moo;
    use strictures 2;;
    use Carp qw(confess);

    has [qw(house senate)] => (
        is => 'ro',
        predicate => 1,
    );

    sub BUILDARGS {
        my ($class, $hr) = @_;

        confess "$class only accepts hashrefs in its constructor" unless ( ref($hr) eq 'HASH' );

        my $out;

        $out->{house} = WebService::Geocodio::Fields::StateLegislativeDistrict::House->new(
            $hr->{house}) if exists $hr->{house};
        $out->{senate} = WebService::Geocodio::Fields::StateLegislativeDistrict::Senate->new(
            $hr->{senate}) if exists $hr->{senate};

        return $out;
    }

}

############### School districts

{
    package WebService::Geocodio::Fields::SchoolDistrict::Unified;

    use Moo;
    use strictures 2;;

    has [qw(name lea_code grade_low grade_high)] => (
        is => 'ro',
    );
}

{
    package WebService::Geocodio::Fields::SchoolDistrict::Elementary;

    use Moo;
    use strictures 2;;

    has [qw(name lea_code grade_low grade_high)] => (
        is => 'ro',
    );
}

{
    package WebService::Geocodio::Fields::SchoolDistrict::Secondary;

    use Moo;
    use strictures 2;;

    has [qw(name lea_code grade_low grade_high)] => (
        is => 'ro',
    );
}

{

    package WebService::Geocodio::Fields::SchoolDistrict;

    use WebService::Geocodio::Fields::SchoolDistrict::Unified;
    use WebService::Geocodio::Fields::SchoolDistrict::Elementary;
    use WebService::Geocodio::Fields::SchoolDistrict::Secondary;

    use Moo;
    use strictures 2;;
    use Carp qw(confess);

    has [qw(unified elementary secondary)] => (
        is => 'ro',
        predicate => 1,
    );

    sub BUILDARGS {
        my ($class, $hr) = @_;

        confess "$class only accepts hashrefs in its constructor" unless ( ref($hr) eq 'HASH' );

        my $out;

        $out->{unified} = WebService::Geocodio::Fields::SchoolDistrict::Unified->new(
            $hr->{unified}) if exists $hr->{unified};
        $out->{elementary} = WebService::Geocodio::Fields::SchoolDistrict::Elementary->new(
            $hr->{elementary}) if exists $hr->{elementary};
        $out->{secondary} = WebService::Geocodio::Fields::SchoolDistrict::Secondary->new(
            $hr->{secondary}) if exists $hr->{secondary};

        return $out;
    }
}

############################ Timezone

{
    package WebService::Geocodio::Fields::Timezone;

    use Moo;
    use strictures 2;

    has [qw(name utc_offset observes_dst)] => (
        is => 'ro',
    );
}

############################ Census

{
    package WebService::Geocodio::Fields::Census;

    use Moo;
    use strictures 2;

    has [qw(state_fips county_fips place_fips tract_code
        block_group block_code census_year)] => (
        is => 'ro',
    );

}

package WebService::Geocodio::Fields;

# ABSTRACT: Modules to represent various fields Geocod.io supports

use WebService::Geocodio::Fields::CongressionalDistrict;
use WebService::Geocodio::Fields::StateLegislativeDistrict;
use WebService::Geocodio::Fields::SchoolDistrict;
use WebService::Geocodio::Fields::Timezone;
use WebService::Geocodio::Fields::Census;

use Moo;
use strictures 2;
use Carp qw(confess);

has [qw(cd stateleg school timezone census)] => (
    is => 'rw',
    predicate => 1,
);

sub BUILDARGS {
    my ( $class, $hr ) = @_;

    confess "$class only accepts hashrefs in its constructor" unless ( ref($hr) eq 'HASH' );

    my $out;

    $out->{cd} = [ map { WebService::Geocodio::Fields::CongressionalDistrict->new($_) }
        @{ $hr->{congressional_districts} } ] if exists $hr->{congressional_districts};
    $out->{stateleg} = WebService::Geocodio::Fields::StateLegislativeDistrict->new(
        $hr->{state_legislative_districts}) if exists $hr->{state_legislative_districts};
    $out->{school} = WebService::Geocodio::Fields::SchoolDistrict->new(
        $hr->{school_districts}) if exists $hr->{school_districts};
    $out->{timezone} = WebService::Geocodio::Fields::Timezone->new($hr->{timezone})
        if exists $hr->{timezone};
    $out->{census} = WebService::Geocodio::Fields::Census->new($hr->{census})
        if exists $hr->{census};

    return $out;
}

1;



lib/WebService/Geocodio/Request.pm
use strict;
use warnings;

package WebService::Geocodio::Request;

use Moo::Role;
use strictures 2;
use HTTP::Tiny;
use Carp qw(confess);
use WebService::Geocodio::Location;

with 'WebService::Geocodio::JSON';

# ABSTRACT: A request role for Geocod.io

=attr ua

A user agent object. Default is L<HTTP::Tiny>

=cut

has 'ua' => (
    is => 'ro',
    lazy => 1,
    default => sub { HTTP::Tiny->new(
        agent => "WebService-Geocodio ",
        default_headers => { 'Content-Type' => 'application/json' },
    ) },
);

=attr base_url

The base url to use when connecting to the service. Default is 'http://api.geocod.io'

=cut

has 'base_url' => (
    is => 'ro',
    lazy => 1,
    default => sub { 'http://api.geocod.io/v1.2/' },
);

=method send_forward

This method POSTs an arrayref of data to the service for processing.  If the
web call is successful, returns an array of L<WebService::Geocodio::Location>
objects.

Any API errors are fatal and reported by C<Carp::confess>.

=cut

sub send_forward {
    my $self = shift;

    $self->_request('geocode', $self->encode(@_));
}

=method send_reverse

This method POSTs an arrayref of data to the service for processing.  If the
web call is successful, returns an array of L<WebService::Geocodio::Location>
objects.

Any API errors are fatal and reported by C<Carp::confess>.

=cut

sub send_reverse {
    my $self = shift;

    $self->_request('reverse', $self->encode(@_));
}

sub _request {
    my ($self, $op, $content) = @_;

    my $url;
    if ( $self->has_fields ) {
        $url = $self->base_url 
            . "$op?fields=" . join(',', @{ $self->fields }) 
            .  "&api_key=" . $self->api_key
            ;
    }
    else {
        $url = $self->base_url . "$op?api_key=" . $self->api_key;
    }

    my $response = $self->ua->request('POST', $url, { content => $content });

    if ( $response->{success} ) {
        my $hr = $self->decode($response->{content});
        return map { WebService::Geocodio::Location->new($_) } 
            map {; @{$_->{response}->{results}} } @{$hr->{results}};
    }
    else {
        confess "Request to " . $self->base_url . "$op failed: (" . 
            $response->{status} . ") - " . $response->{content};
    }
}


1;


lib/WebService/Geocodio/Location.pm
use strict;
use warnings;

package WebService::Geocodio::Location;

use WebService::Geocodio::Fields;
use Moo;
use strictures 2;
use Carp qw(confess);

# ABSTRACT: Location object for use with Geocod.io service.

=attr number

Buiding number

=attr street

A street name identifier

=attr postdirection

A address direction like 'SW' or 'S'

=attr suffix

The type of street 'Ave', 'St', etc.

=attr city

The city where the streets have no name.

=attr state

State wherein city is located

=attr zip

The postal zip code

You B<must> have either a zip code OR a city/state pair.

=attr lat

The latitude of the location

=attr lng

The longitude of the location

=attr accuracy

A float from 0 -> 1 representing the confidence of the lookup results

=attr formatted

The full address as formatted by the service.

=attr fields

Any requested fields are available here.

=cut

has [qw(number street suffix postdirection city state zip formatted lat lng accuracy fields)] => (
    is => 'ro',
    predicate => 1,
);

=method new

The constructor accepts either a bare string OR a list of key/value pairs where
the keys are the attribute names.

=cut

sub BUILDARGS {
    my ( $class, @args ) = @_;

    my $out;
    if (ref($args[0]) eq "HASH") {
        my $hr = $args[0];
        $out->{accuracy} = $hr->{accuracy} if exists $hr->{accuracy};
        $out->{formatted} = $hr->{formatted_address} if exists $hr->{formatted_address};
        $out->{fields} = WebService::Geocodio::Fields->new($hr->{fields}) if exists $hr->{fields};
        map { $out->{$_} = $hr->{address_components}->{$_} if exists $hr->{address_components}->{$_} } qw(number street suffix postdirection city state zip);
        map { $out->{$_} = $hr->{location}->{$_} if exists $hr->{location}->{$_} } qw(lat lng);
    }
    elsif ( @args % 2 == 1 ) {
        $out->{formatted} = $args[0];
    }
    else {
        $out = { @args };
    }

    return $out;
}

sub _forward_formatting {
    my $self = shift;

    return $self->formatted if $self->has_formatted;

    if ( ( not $self->has_zip ) && ( not ( $self->has_city && $self->has_state ) ) ) {
        confess "A zip or city-state pair is required.\n";
    }

    my $s;
    if ( $self->has_number && $self->has_street && $self->suffix ) {
        my @f;
        if ( $self->has_postdirection ) {
            @f = qw(number postdirection street suffix);
        }
        else {
            @f = qw(number street suffix);
        }

        $s .= join " ", (map {; $self->$_ } @f);
        $s .= ", ";
    }

    if ( $self->has_city && $self->has_state) {
        $s .= join ", ", (map {; $self->$_ } qw(city state));
        $s .= " " if ( $self->has_zip );
    }

    if ( $self->has_zip ) {
        $s .= $self->zip
    }

    return $s;
}

sub _reverse_formatting {
    my $self = shift;

    if ( not ( $self->has_lat && $self->has_lng ) ) {
        confess "lat-lng pair is required\n";
    }

    return join ",", ( map {; $self->$_ } qw(lat lng) );
}


1;


lib/WebService/Geocodio.pm
use 5.014;

use strict;
use warnings;

package WebService::Geocodio;

use Moo;
use strictures 2;
use Carp qw(confess);
use Scalar::Util qw(blessed);
with('WebService::Geocodio::Request');

# ABSTRACT: A Perl interface to Geocod.io

=head1 SYNOPSIS

    use 5.014;
    use WebService::Geocodio;
    use WebService::Geocodio::Location;

    my $geo = WebService::Geocodio->new(
        api_key => $ENV{GEOCODIO_API_KEY}
    );

    # Wrigley Field
    my $loc = WebService::Geocodio::Location->new(
        number => 1060,
        postdirection => 'W',
        street => 'Addison',
        suffix => 'Street',
        city => 'Chicago',
        state => 'IL',
    );

    # Could add more than one thing here, even bare strings are OK
    # 20050 = zip code in Washington DC
    $geo->add_location($loc, '20050');

    $geo->add_field('timezone');

    # prints:
    # Chicago: 41.947205791667, -87.656316875, CST
    # Chicago: 41.947180613636, -87.657167363636, CST
    # Washington: 38.893311, -77.014647, EST
    map { say $_->city, ": ", $_->lat, ", ", $_->lng, ", " $_->fields->timezone->name } $geo->geocode();

=head1 OVERVIEW

This module is a fairly thin wrapper around the L<Geocod.io|http://geocod.io>
geocoding web service.  This service currently supports US and many Canadian based
addresses at the moment.  Both forward and reverse geocoding is supported.

In my testing, the service is somewhat finicky about how addresses are
presented and stored; please read the service API documentation thoroughly 
to make sure you're getting the best quality results from the service.

You will need to obtain a free API key to use this library.

All errors are fatal and reported by C<confess>.  If you want more graceful
error handling, you might want to try using L<Try::Tiny>.

=cut

=attr api_key

This is the geocod.io API key. It is required.

=cut

has 'api_key' => (
    is => 'ro',
    isa => sub { confess "$_[0] doesn't look like a valid api key\n" unless $_[0] =~ /[0-9a-f]+/ },
    required => 1,
);

=attr locations

The list of locations you want to geocode.  These can be bare strings (if you like) or
you can use a fancy object like L<WebService::Geocodio::Location> which will serialize
itself to JSON automatically.

=cut

has 'locations' => (
    is => 'rw',
    default => sub { [] },
);

=attr fields

You may request the following fields be included in the results:

=over 4

=item * cd

Congressional District (for the current Congress). This will return detailed
biographical information about the Senators and Congressional representative
from this district in the C<current_legislators> field which is structured
as an array where the member of Congress is first, followed by the Senators
for the entire state.

It's possible there may be overlaps in a given location, so this field is
structured as an array since there may be 2 or more possible districts
for a given location. See the official docs for more information.

=item * cd113

Congressional District (for the 113th Congress which ran through 2015)

=item * cd114

Congressional District (for the 114th Congress which ran through 2017)

=item * cd115

Congressional District (for the 115th Congress which runs through 2019)

=item * stateleg

The state legislative divisions for this location. The results include both
House and Senate, unless the location is unicameral like Nebraska or Washington
D.C., then only a senate result is given.

=item * timezone

The timezone of this location, UTC offset and whether it observes daylight
saving time.

=item * school

The unified or elementary/secondary school district identifiers for this location.

=item * census

Appends various U.S. Census statistical codes to your lookup. See the API docs
for more information about these codes and how they can be used with other
census data sets.

=back

=cut

has 'fields' => (
    is => 'rw',
    predicate => 1,
    default => sub { [] },
);

=method add_location

This method takes one or more locations and stores it in the locations attribute.

=cut

sub add_location {
    my $self = shift;

    push @{ $self->locations }, @_;
}

=method show_locations

Show the locations currently set for geocoding.

=cut

sub show_locations {
    my $self = shift;

    return @{ $self->locations };
}

=method clear_locations

If you want to clear the current list of locations, use this method.

=cut

sub clear_locations {
    my $self = shift;

    $self->locations([]);
}

=method add_field

This method takes one or more fields to include in a result set. Valid fields are:

=over 4

=item * cd (current Congress)

=item * cd113

=item * cd114

=item * cd115

=item * stateleg

=item * timezone

=item * school

=item * census

=back

Fields that do not match these valid names are silently discarded.

=cut

sub add_field {
    my $self = shift;

    push @{ $self->fields }, grep { /cd|cd113|cd114|cd115|stateleg|timezone|school|census/ } @_;
}

=method geocode

Send the current list of locations to the geocod.io service.

Returns undef if there are no locations stored.

In a list context, returns a list of L<WebService::Geocodio::Location> objects.
In a scalar context, returns an arrayref of L<WebService::Geocodio::Location>
objects. The list of objects is presented in descending order of accuracy.

=cut

sub geocode {
    my $self = shift;

    $self->add_location(@_) if scalar @_;

    return undef if scalar @{$self->locations} < 1;

    my @r = $self->send_forward( $self->_format('forward') );

    wantarray ? return @r : return \@r;
}

=method reverse_geocode

Send the current list of latitude, longitude pairs to the geocod.io service.

Returns undef if there are no locations stored.

In a list context, returns a list of L<WebService::Geocodio::Location> objects.
In scalar context, returns an arrayref of L<WebService::Geocodio::Location> 
objects.  The list of objects is presented in descending order of accuracy.

=cut

sub reverse_geocode {
    my $self = shift;

    $self->add_location(@_) if scalar @_;

    return undef if scalar @{$self->locations} < 1;

    my @r = $self->send_reverse( $self->_format('reverse') );

    wantarray ? return @r : return \@r;
}

sub _format {
    my $self = shift;
    my $direction = shift;

    my $method = $direction eq 'forward' ? '_forward_formatting'
        : '_reverse_formatting';

    return [ map {; blessed $_ ? $_->$method : $_ } @{$self->locations} ];
}

1;
